pkg/private/build_tar.py

```diff
@@ -197,6 +197,91 @@ def add_deb(self, deb):
       self.add_tar(tmpfile[1])
       os.remove(tmpfile[1])
 
+  def add_tree(self, tree_top, destpath, mode=None, ids=None, names=None):
+    """Add a tree artifact to the tar file.
+
+    Args:
+       tree_top: the top of the tree to add
+       destpath: the path under which to place the files
+       mode: force to set the specified mode, by default the value from the
+         source is taken.
+       ids: (uid, gid) for the file to set ownership
+       names: (username, groupname) for the file to set ownership. `f` will be
+         copied to `self.directory/destfile` in the layer.
+    """
+    dest = destpath.strip('/')  # redundant, dests should never have / here
+    if self.directory and self.directory != '/':
+      dest = self.directory.lstrip('/') + '/' + dest
+
+    dest = os.path.normpath(dest).replace(os.path.sep, '/')
+    # If mode is unspecified, derive the mode from the file's mode.
+    if mode is None:
+      mode = 0o755 if os.access(f, os.X_OK) else 0o644
+    if ids is None:
+      ids = (0, 0)
+    if names is None:
+      names = ('', '')
+
+    to_write = {}
+    for root, dirs, files in os.walk(tree_top):
+      dirs = sorted(dirs)
+      rel_path_from_top = root[len(tree_top):].lstrip('/')
+      if rel_path_from_top:
+        dest_dir = dest + '/' + rel_path_from_top + '/'
+      else:
+        dest_dir = dest + '/'
+      for dir in dirs:
+        to_write[dest_dir + dir] = None
+      for file in sorted(files):
+        to_write[dest_dir + file] = os.path.join(root, file)
+
+    for path in sorted(to_write.keys()):
+      content_path = to_write[path]
+      if not content_path:
+        self.add_empty_file(
+            path,
+            mode=mode,
+            ids=ids,
+            names=names,
+            kind=tarfile.DIRTYPE)
+      else:
+        self.tarfile.add_file(
+            path,
+            file_content=content_path,
+            mode=mode,
+            uid=ids[0],
+            gid=ids[1],
+            uname=names[0],
+            gname=names[1])
+
+  def add_manifest_entry(self, entry, file_attributes):
+    dest, src, mode, user, group, link, is_dir = entry
+
+    # Use the pkg_tar mode/owner remaping as a fallback
+    non_abs_path = dest.strip('/')
+    if file_attributes:
+      attrs = file_attributes(non_abs_path)
+    else:
+      attrs = {}
+    # But any attributes from the manifest have higher precedence
+    if mode is not None and mode != '':
+      attrs['mode'] = int(mode, 8)
+    if user:
+      if group:
+        attrs['names'] = (user, group)
+      else:
+        # Use group that legacy tar process would assign
+        attrs['names'] = (user, attrs.get('names')[1])
+    if link:
+      self.add_link(dest, link)
+    elif is_dir == 1:
```

Comment:
@aiuto once options have [descriptive names](https://github.com/bazelbuild/rules_pkg/pull/367/files#r652544583), this bit would become more readable too

Reply:
Switched to named constants.  PTAL

Reply:
LGTM.